Documentation/design-docs/arm64-intrinsics.md

```diff
@@ -0,0 +1,217 @@
+
+`IsSupported` will be used in any namespace which may not be implemented on all CoreCLR platforms
+
+### Intrinsic Interface Documentation
+
++ Documentation will be minimal preferring the underlying ARM documentation
++ Namespaces will briefly document corresponding specification
++ Intrinsic methods will briefly document corresponding assembly instruction(s) for each platform
+
+### Test coverage
+
+As intrinsic support is added test coverage must be extended to provide basic testing
+
+### Phased Implementation
+
+#### Implementation Priorities
+
+As guidelines:
+
++ Baseline functionality will be prioritized over architectural extensions
++ Architectural extensions will typically be prioritized in age order.  Earlier extensions will be added first
++ `ARM64` only and `ARM*` features will have equal priority
++ Priorities will be driven by optimization efforts and requests
++ Priority will be given to intrinsics which have already been implemented for other platforms
++ Priority will be given to intrinsics which are equivalent/similar to those actively used in libraries for other platforms
+
+#### Addition of new intrinsics namespaces
+When an intrinsic namespace is introduced code will likely be implemented on different platforms at different times.
+
+It is explicitly OK for ARM64 to implement a namespace which may be implemented on `ARM32` w/o `ARM32` implementation.
+The opposite is also true.
+
+Any platform which does not implement a namespace, must:
+
++ Return `false` for `IsSupported`
++ `throw` on any unimplemented function
+
+#### Partial implementation of intrinsic namespace
+
++ It is preferred that `IsSupported` represents the state of an entire namespace
++ It is certainly required at time of a release, `IsSupported` represents the state of an entire namespace
++ Allow for writing tests and implementing code, `IsSupported == false` may be used when namespace is partially implemented
+
+#### Addition of new intrinsics to existing namespace
+When an intrinsic namespace is introduced, a best guess will be made about the set of useful intrinsics.
+However, the intrinsics will be added as needed or requested for various optimization efforts.
+
++ The set of intrinsics instructions supported by a namespace should only be allowed
+if all platforms reporting `IsSupported` are supported (at least at time of release)
++ It may be preferable to add new namespaces in cases where this creates significant burden
+
+### Half precision floating point
+
+This document will refer to half precision floating point as `Half`.
+
++ Machine learning and Artificial intelligence often use `Half` type to simplify storage and improve processing time.
++ CoreCLR and `CIL` in general do not have general support for a `Half` type
++ There is an open request to expose `Half` intrinsics
++ ARM64 will introduce `System.Runtime.Intrinsics.Half` to support this request 
```

Comment:
IMO it is wrong. Half and other binary/floating ISO 854 and non-ISO types should be available at the System namespace. We may need 8bit floating types as well.

Reply:
Long term, I think ECMA specs need to be amended to add these data types.

Short term, use of a struct as an opaque type to represent Half seems reasonable.

Given the conflict between long term goals and short term needs.  It seemed best to keep it in a limited namespace.  However, it can be moved anywhere.